api/client/client_test.go

```diff
@@ -0,0 +1,487 @@
+		t.Fatal("model found by id does't equal the original")
+	}
+}
+
+func TestReadTransaction(t *testing.T) {
+	_, clean := server(t)
+	defer clean()
+	client := client(t)
+
+	storeID, err := client.NewStore()
+	checkErr(t, err)
+	err = client.RegisterSchema(storeID, modelName, schema)
+	checkErr(t, err)
+	err = client.Start(storeID)
+	checkErr(t, err)
+	person := createPerson()
+	err = client.ModelCreate(storeID, modelName, person)
+	checkErr(t, err)
+
+	txn, err := client.ReadTransaction(storeID, modelName)
+	if err != nil {
+		t.Fatalf("failed to create read txn: %v", err)
+	}
+
+	defer func() {
+		err = txn.End()
+		if err != nil {
+			t.Fatalf("failed to end txn: %v", err)
+		}
+	}()
+
+	err = txn.Start()
+	if err != nil {
+		t.Fatalf("failed to start read txn: %v", err)
+	}
+
+	has, err := txn.Has(person.ID)
+	if err != nil {
+		t.Fatalf("failed to read txn has: %v", err)
+	}
+	if !has {
+		t.Fatal("expected has to be true but it wasn't")
+	}
+
+	foundPerson := &Person{}
+	err = txn.FindByID(person.ID, foundPerson)
+	if err != nil {
+		t.Fatalf("failed to txn find by id: %v", err)
+	}
+	if !reflect.DeepEqual(foundPerson, person) {
+		t.Fatal("txn model found by id does't equal the original")
+	}
+}
+
+func TestWriteTransaction(t *testing.T) {
+	_, clean := server(t)
+	defer clean()
+	client := client(t)
+
+	storeID, err := client.NewStore()
+	checkErr(t, err)
+	err = client.RegisterSchema(storeID, modelName, schema)
+	checkErr(t, err)
+	err = client.Start(storeID)
+	checkErr(t, err)
+	existingPerson := createPerson()
+	err = client.ModelCreate(storeID, modelName, existingPerson)
+	checkErr(t, err)
+
+	txn, err := client.WriteTransaction(storeID, modelName)
+	if err != nil {
+		t.Fatalf("failed to create write txn: %v", err)
+	}
+
+	defer func() {
+		err = txn.End()
+		if err != nil {
+			t.Fatalf("failed to end txn: %v", err)
+		}
+	}()
+
+	err = txn.Start()
+	if err != nil {
+		t.Fatalf("failed to start write txn: %v", err)
+	}
+
+	person := createPerson()
+
+	err = txn.Create(person)
+	if err != nil {
+		t.Fatalf("failed to create in write txn: %v", err)
+	}
+	if person.ID == "" {
+		t.Fatalf("expected an entity id to be set but it wasn't")
+	}
+
+	has, err := txn.Has(existingPerson.ID)
+	if err != nil {
+		t.Fatalf("failed to write txn has: %v", err)
+	}
+	if !has {
+		t.Fatalf("expected has to be true but it wasn't")
+	}
+
+	foundExistingPerson := &Person{}
+	err = txn.FindByID(existingPerson.ID, foundExistingPerson)
+	if err != nil {
+		t.Fatalf("failed to txn find by id: %v", err)
+	}
+	if !reflect.DeepEqual(foundExistingPerson, existingPerson) {
+		t.Fatalf("txn model found by id does't equal the original")
+	}
+
+	existingPerson.Age = 99
+	err = txn.Save(existingPerson)
+	if err != nil {
+		t.Fatalf("failed to save in write txn: %v", err)
+	}
+
+	err = txn.Delete(existingPerson.ID)
+	if err != nil {
+		t.Fatalf("failed to delete in write txn: %v", err)
+	}
+}
+
+func TestListen(t *testing.T) {
```

Comment:
This test hangs forever and times out.

Reply:
I'll take a look after the review.

Reply:
It does in JS as well. Is there an API to 'cancel' the stream/listener via the gRPC API?

Reply:
To leave the note here. This problem was solved in a further commit in this PR. The server-side wasn't checking if the gRPC call context was cancelled, thus stayed in an infinite loop which made the `GracefulShutdown` of the server in the test to hang forever.

Reply:
Yep, fixed.